#1586  Equivalence check for sets

ueliwechsler opened the pull request:
Close #1586

mforets commented on src/Interfaces/LazySet.jl:
````diff
@@ -682,3 +683,34 @@ end
+```
+"""
+function isequivalent(X::LazySet, Y::LazySet)
+    if X == Y
````
Ok; this check is useful if the sets are of the same type, that was the case?

ueliwechsler replied:
Yes, I added this because it is not expensive to check `X == Y` and could speed up the check if `X == Y` is already true.

mforets replied:
Ok, makes sense.

mforets commented on src/Interfaces/LazySet.jl:
````diff
@@ -682,3 +683,34 @@ end
+julia> is_equivalent(X, Y)
+true
+```
+"""
````
Let's add the algorithm:

```julia
## Algorithm

First, the check `X == Y` is performed which returns `true` if and only if the given sets are of the same type, and have the same values (modulo floating-point tolerance). Otherwise, the double inclusion check `X ⊆ Y && Y ⊆ X` is computed.
```

ueliwechsler commented:
Do you guys know, what is wrong with the build travis build in julia 1.2?

mforets commented:
It says `GitError(Code:ERROR, Class:Net, curl error: Failed to connect to github.com port 443: Connection timed out`, i will try triggering the build again.

schillic commented:
The v1.2 builds run into timeouts. Not sure why.

schillic commented on src/Interfaces/LazySet.jl:
```diff
@@ -682,3 +683,38 @@ end
+julia> Y = convert(HPolytope, X)
+julia> X == Y
+false
+julia> is_equivalent(X, Y)
```
```suggestion
julia> isequivalent(X, Y)
```

schillic commented on test/unit_util.jl:
```diff
@@ -85,4 +86,12 @@ for _dummy_ in 1:1 # avoid global variable warnings
+    I = Diagonal(ones(n))
+    X = HPolytope([I; -I], ones(2*n))
+    Y = convert(VPolytope, X)
+    @test LazySets.isequivalent(X, Y)
```
```suggestion
    @test isequivalent(X, Y)
```

schillic commented on src/Interfaces/LazySet.jl:
````diff
@@ -682,3 +683,38 @@ end
+
+### Examples
+
+```
````
```suggestion
```jldoctest
```
This would detect the other error I reported above. However, you need to add blank lines after each command, like so:
```julia
julia> X = BallInf([0.1, 0.2], 0.3);

julia> Y = convert(HPolytope, X);

julia> X == Y
false

julia> isequivalent(X, Y)
true
```

ueliwechsler commented:
> The v1.2 builds run into timeouts. Not sure why.

Is this a general problem at the moment. Or does it have to do something with my code?

schillic commented:
> Is this a general problem at the moment.

I think so because it does not even pass the first test.

ueliwechsler commented:
Do you have any idea, what I could change to make it pass the tests?

schillic commented:
I actually tried and see the same result in my v1.2 version, so it is not an issue with the build server. I do not have time to debug now, but maybe you want to try running the tests locally in v1.2 and see why they do not pass your tests. If I had to guess, I would say that there is something with `Rational`s for your programs that does not terminate.

ueliwechsler commented:
Thanks, I will try it later on.

schillic commented:
Out of curiosity: What was the problem?

ueliwechsler commented:
> Out of curiosity: What was the problem?

I am not sure. I didn't had the motivation to see why it did not work, once it worked 😆. I changed two main things. First, I added the test at a different location `unit_polytopes.jl` and I removed the use of `Diagonal` and therefore the import of `LinearAlgebra: Diagonal`.

mforets commented:
The problem was somehow related to polyhedra dependency. For instance, leaving the test in `unit_util.jl` but changing it to

```julia
    # Check if sets are equal by double inclusion
    n = 2
    I = Diagonal(ones(n))
    X = HPolygon([I; -I], ones(2*n))
    Y = convert(VPolygon, X)
    @test isequivalent(X, Y)
```
worked for me. (I just didn't have time to communicate it yesterday!)

mforets commented:
Merge?

Cc @schillic

schillic commented on src/Interfaces/LazySet.jl:
````diff
@@ -682,3 +683,44 @@ end
+```
+"""
+function isequivalent(X::LazySet, Y::LazySet)
+    if X == Y
````
Since the inclusion checks are approximate, it could be more consistent to use `X ≈ Y` here. At least this would match the documentation. (So at least the documentation or the code have to be changed here.)
However, `≈` is not controlled by our internal precision variables, so it might also be confusing. Thoughts?

ueliwechsler replied:
I would suggest adding `≈`, then it fits the documentation.

If you want to make it less confusing, we could also just remove this preceding equality check.

mforets replied:
> I would suggest adding ≈, then it fits the documentation.

I think this is as good idea :thumbsup:

> If you want to make it less confusing, we could also just remove this preceding equality check.

We shouldn't remove this initial comparison: if the sets are of the same type, it is going to be more efficient than the double inclusion check.

> However, ≈ is not controlled by our internal precision variables

Good point, that should be fixed by using `_isapprox` in line 437.

mforets replied:
> Good point, that should be fixed by using _isapprox in line 437.

This is now https://github.com/JuliaReach/LazySets.jl/pull/1824

schillic commented on test/unit_Polytope.jl:
```diff
@@ -228,6 +228,10 @@ for N in [Float64]
 
+    # double inclusion check
+    Z = convert(VPolytope, Y)
+    @test isequivalent(Y, Z)
```
Can you add a negative test as well?

ueliwechsler replied:
Good enough?

schillic replied:
Thanks!

schillic commented on src/Interfaces/LazySet.jl:
```diff
@@ -682,3 +683,44 @@ end
+
+First, the check `X == Y` is performed which returns `true` if and only if the given sets are of the same type,
+and have the same values (modulo floating-point tolerance). Otherwise, the double inclusion check `X ⊆ Y && Y ⊆ X` is
+computed.
```
```suggestion
used.
```
I would say that a test is not computed.

mforets commented:
Thanks a lot!
